### ISSUE TITLE
Server: GridServiceHealthMonitorJob can pile up deployments

### ISSUE THREAD

**Issue by jnummelin** [kontena/kontena#2208]
Given a failing health check, `GridServiceHealthMonitorJob` will just blindly create new deployments if the service overall health drops too much. In case of mis-configured (or anyway heavily failing) health check this can cause service deployments to pile up heavily with these kind of log entries
```
I, [2017-04-24T11:48:10.369721 #11]  INFO -- GridServiceSchedulerWorker: delaying foo/bar/some-service deploy because there is another deploy in progress
```
These will probably also mean timeouts with distributed locks:
```
W, [2017-04-24T11:52:45.536157 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:45.763583 #9]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:45.794109 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:45.805272 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:46.810364 #9]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:46.812494 #9]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:46.845791 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:56.563669 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:56.793793 #9]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:56.815847 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
W, [2017-04-24T11:52:56.844696 #11]  WARN -- Module: timeout after waiting 10.0s of 10.0s until: lock check_deploy_queue:580ddcc376f5980007001319 is available
```

`GridServiceHealthMonitorJob` should check if there is already a pending deployment for the given service and only if not create the deployment.

**Comment by SpComb** [comment]
The `GridServiceHealthMonitorJob` should not consider an unhealthy service as needing deployment if it has pending deploys: https://github.com/kontena/kontena/blob/v1.2.0/server/app/jobs/grid_service_health_monitor_job.rb#L33

```ruby
 def deploy_needed?(service)
    ...
    if health_percent < expected_health
      service.running? && !service.deploy_pending?
    else
      false
    end
  end
```

Is this logic broken somehow?

**Comment by jnummelin** [comment]
possibly yes, `GridService#deploy_pending?` only checks if there are deploys with `started_at=nil`. Now when the `GridServiceSchedulerWorker` check the deploy "queue", it marks deploys with some timestamp for `started_at` when it picks them up from the queue and only after a 30 "wait" marks them back at not started if there were previous ongoing deploys.

**Comment by SpComb** [comment]
> I think it should rather check finished_at being nil, not sure if that would have any side-effects though.

That sounds saner to me.

**Comment by SpComb** [comment]
Fixing the `GridService#deploy_pending?` to avoid the `GridServiceHealthMonitorJob` spam is a good idea, but it seems like that kind of deploy spam also exposes another bug in the `GridServiceSchedulerWorker`... the `deploying?` check also picks up any other pending deploys during that 30s. This causes all of the `GridServiceDeploy`s for a `GridService` to get stuck, and the service will not get deployed anymore...

---

A `GridService` is deployed:

```
D, [2017-04-24T18:01:19.219552 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:19.223824 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
```

Some `GridServiceSchedulerWorker` picks up the `GridServiceDeploy`, marks it as `started_at`, takes the `check_deploy_queue` dlock for the `GridService`, and starts performing the deploy:

```
I, [2017-04-24T18:01:19.342830 #14485]  INFO -- GridServiceDeployer: starting to deploy development/null/redis-unhealthy
I, [2017-04-24T18:01:19.453981 #14485]  INFO -- GridServiceInstanceDeployer: Deploying service instance development/null/redis-unhealthy-1 to node core-01 at 2017-04-24 15:01:19 UTC...
D, [2017-04-24T18:01:19.986539 #14485] DEBUG -- GridServiceInstanceDeployer: waiting 1.0s of 300.0s until: service development/null/redis-unhealthy-1 is running on node development/core-01 at 2017-04-24 15:01:19 UTC
```

Meanwhile, before the deploy completes, the same `GridService` gets re-deployed:

```
D, [2017-04-24T18:01:20.287823 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:20.290590 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
```

Some other `GridServiceSchedulerWorker`s pick up that `GridServiceDeploy`, marks it as `started_at`, and then attempts to take the `check_deploy_queue` dlock:

```
D, [2017-04-24T18:01:21.065715 #14489] DEBUG -- Module: waiting 0.6s of 10.0s until: lock check_deploy_queue:58fe08b9de35785cd200000a is available
```

Meanwhile, even more deploys get queued:

```
D, [2017-04-24T18:01:23.482476 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:23.486851 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
```

```
I, [2017-04-24T18:01:23.494959 #14485]  INFO -- GridServiceInstanceDeployer: waited 4.0s of 300.0s until: service development/null/redis-unhealthy-1 is running on node development/core-01 at 2017-04-24 15:01:19 UTC yielded GridServiceInstance
```

```
D, [2017-04-24T18:01:25.130112 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:25.134306 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
```

Once the initial service deploy completes:

```
I, [2017-04-24T18:01:29.244643 #14485]  INFO -- GridServiceDeployer: service development/null/redis-unhealthy has been deployed
```

Some other `GridServiceSchedulerWorker` gets the `check_deploy_queue` dlock for the service:

```
 I, [2017-04-24T18:01:29.264757 #14489]  INFO -- Module: waited 8.7s of 10.0s until: lock check_deploy_queue:58fe08b9de35785cd200000a is available yielded String
```

And then it calls `service_deploy.grid_service.deploying?`... but this picks up all of the other pending `GridServiceDeploy`s that have been picked up by the other workers, each of them having `started_at` set:

```
I, [2017-04-24T18:01:29.272482 #14489]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
```

This happens for each of the pending deploys:

```
I, [2017-04-24T18:01:29.281760 #14489]  INFO -- Module: waited 5.7s of 10.0s until: lock check_deploy_queue:58fe08b9de35785cd200000a is available yielded String
I, [2017-04-24T18:01:29.314103 #14489]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
I, [2017-04-24T18:01:29.360172 #14493]  INFO -- Module: waited 4.2s of 10.0s until: lock check_deploy_queue:58fe08b9de35785cd200000a is available yielded String
I, [2017-04-24T18:01:29.385816 #14493]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
```

At this point, the service is screwed, and it won't get deployed anymore:

```
D, [2017-04-24T18:01:30.137231 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:30.139887 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
I, [2017-04-24T18:01:30.203485 #14493]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
D, [2017-04-24T18:01:31.330360 #14485] DEBUG -- GridServiceHealthMonitorJob: service=redis-unhealthy health=0.0 deploy_pending=false
I, [2017-04-24T18:01:31.333296 #14485]  INFO -- GridServiceHealthMonitorJob: service health too low, triggering full deploy for development/null/redis-unhealthy
I, [2017-04-24T18:01:31.410615 #14493]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
```

With a sufficient number of overlapping service deploys, even after the `GridServiceSchedulerWorker`s `after(30) { service_deploy.set(:started_at => nil) }` happens, it doesn't help:

```
D, [2017-04-24T18:01:59.275154 #14489] DEBUG -- GridServiceSchedulerWorker: resetting delayed development/null/redis-unhealthy deploy
I, [2017-04-24T18:01:59.297971 #14493]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
D, [2017-04-24T18:01:59.314599 #14489] DEBUG -- GridServiceSchedulerWorker: resetting delayed development/null/redis-unhealthy deploy
I, [2017-04-24T18:01:59.329628 #14485]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
D, [2017-04-24T18:01:59.386733 #14493] DEBUG -- GridServiceSchedulerWorker: resetting delayed development/null/redis-unhealthy deploy
I, [2017-04-24T18:01:59.420794 #14489]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
D, [2017-04-24T18:02:00.203921 #14493] DEBUG -- GridServiceSchedulerWorker: resetting delayed development/null/redis-unhealthy deploy
I, [2017-04-24T18:02:00.289977 #14493]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
D, [2017-04-24T18:02:01.411922 #14493] DEBUG -- GridServiceSchedulerWorker: resetting delayed development/null/redis-unhealthy deploy
I, [2017-04-24T18:02:01.453983 #14489]  INFO -- GridServiceSchedulerWorker: delaying development/null/redis-unhealthy deploy because there is another deploy in progress
```

Because at this point there is a large number of pending `GridServiceDeploy`s, at least some of which are going to have `started_at` set whenever any deploy wakes up...

**Comment by SpComb** [comment]
Created issues for the two `GridServiceSchedulerWorker` bugs revealed by this case, which cause service deploys to get stuck:

*  Service deploys can get stuck if there are several pending deploys for a service that is already being deployed #2212 
*  Service deploys can get stuck if there is a pending deploy while a service deploy is running for more than 10 seconds #2213

**Comment by SpComb** [comment]
Pending the fixes in #2221, if this happens and it triggers a storm of pending deploys, workaround is to `kontena service stop`, and let them clear away.
